notebooks/asian_barrier_option/mc_pricing.ipynb

```diff
@@ -0,0 +1,884 @@
+    "number_of_blocks = (N_PATHS-1) // number_of_threads + 1\n",
+    "output = cupy.zeros(N_PATHS, dtype=cupy.float32)\n",
+    "shared_buffer_size = number_of_threads * 4\n",
+    "numba_gpu_barrier_option_shared_mem[(number_of_blocks,), (number_of_threads,), 0, shared_buffer_size](output, np.float32(T), np.float32(K), \n",
+    "                    np.float32(B), np.float32(S0), \n",
+    "                    np.float32(sigma), np.float32(mu), \n",
+    "                    np.float32(r), randoms_gpu, N_STEPS, N_PATHS)\n",
+    "s = time.time()\n",
+    "numba_gpu_barrier_option_shared_mem[(number_of_blocks,), (number_of_threads,), 0, shared_buffer_size](output, np.float32(T), np.float32(K), \n",
+    "                    np.float32(B), np.float32(S0), \n",
+    "                    np.float32(sigma), np.float32(mu), \n",
+    "                    np.float32(r), randoms_gpu, N_STEPS, N_PATHS)\n",
+    "v = output.mean()\n",
+    "cuda.synchronize()\n",
+    "e = time.time()\n",
+    "print('time', e-s, 'v', v)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### CUPY GPU\n",
+    "CuPy provides an easy way to define GPU kernels from raw CUDA source. `RawKernel` object allows you to call the kernel with CUDA’s `cuLaunchKernel` interface. Here is an example where we wrap the Barrier Option computation code inside the `RawKernel`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cupy_barrier_option = cupy.RawKernel(r'''\n",
+    "extern \"C\" __global__ void barrier_option(\n",
+    "    float *d_s,\n",
+    "    const float T,\n",
+    "    const float K,\n",
+    "    const float B,\n",
+    "    const float S0,\n",
+    "    const float sigma,\n",
+    "    const float mu,\n",
+    "    const float r,\n",
+    "    const float * d_normals,\n",
+    "    const long N_STEPS,\n",
+    "    const long N_PATHS)\n",
+    "{\n",
+    "  unsigned idx =  threadIdx.x + blockIdx.x * blockDim.x;\n",
+    "  unsigned stride = blockDim.x * gridDim.x;\n",
+    "  unsigned tid = threadIdx.x;\n",
+    "\n",
+    "  const float tmp1 = mu*T/N_STEPS;\n",
+    "  const float tmp2 = exp(-r*T);\n",
+    "  const float tmp3 = sqrt(T/N_STEPS);\n",
+    "  double running_average = 0.0;\n",
+    "\n",
+    "  for (unsigned i = idx; i<N_PATHS; i+=stride)\n",
+    "  {\n",
+    "    float s_curr = S0;\n",
+    "    unsigned n=0;\n",
+    "    for(unsigned n = 0; n < N_STEPS; n++){\n",
+    "       s_curr += tmp1 * s_curr + sigma*s_curr*tmp3*d_normals[i + n * N_PATHS];\n",
+    "       running_average += (s_curr - running_average) / (n + 1.0) ;\n",
+    "       if (running_average <= B){\n",
+    "           break;\n",
+    "       }\n",
+    "    }\n",
+    "\n",
+    "    float payoff = (running_average>K ? running_average-K : 0.f);\n",
+    "    d_s[i] = tmp2 * payoff;\n",
+    "  }\n",
+    "}\n",
+    "\n",
+    "''', 'barrier_option')"
```

Comment:
the 16G card cannot run 8m paths distributed MC simulation.  All my deep learning results are based on ground truth calculated from 8m paths. I think we can mention running this in small memory GPU need to adjust down the number of paths

Reply:
This was not related to 16G cards. This is the bug that you were note able to reproduce in the Prometheus cluster

In my case I think the dask workers stalled because of cupy kernel caching inconsistencies; but these cupy kernel caching inconsistencies is due to my particular setup because I map user (myself) and home directory into the container for development/testing purposes. It's a container setup corner case as typically the containers are run without mapping an actual user and his/her home directory into the container, so then they won't see this issue.

I'll mark resolved since it's my specific setup that seems to be the issue.